strawberryfields/configuration.py

```
@@ -0,0 +1,189 @@
+
+.. currentmodule:: strawberryfields.configuration.Configuration
+
+.. autosummary::
+    path
+    load
+    save
+
+Code details
+~~~~~~~~~~~~
+
+.. currentmodule:: strawberryfields.configuration
+
+"""
+import os
+import logging as log
+
+import toml
+from appdirs import user_config_dir
+
+log.getLogger()
+
+
+DEFAULT_CONFIG = {"api": {"authentatication_token": "", "hostname": "localhost", "use_ssl": True}}
+
+
+class ConfigurationError(Exception):
+    """Exception used for configuration errors"""
+
+
+class Configuration:
+    """Configuration class.
+
+    This class is responsible for loading, saving, and storing StrawberryFields
+    and plugin/device configurations.
+
+    Args:
+        name (str): filename of the configuration file.
+        This should be a valid TOML file. You may also pass an absolute
+        or a relative file path to the configuration file.
+    """
+
+    def __str__(self):
+        return "{}".format(self._config)
+
+    def __repr__(self):
+        return "Strawberry Fields Configuration <{}>".format(self._filepath)
+
+    def __init__(self, name="config.toml"):
+        # Look for an existing configuration file
+        self._config = DEFAULT_CONFIG
+        self._config_file = {}
+        self._filepath = None
+        self._name = name
+        self._user_config_dir = user_config_dir("strawberryfields", "Xanadu")
+        self._env_config_dir = os.environ.get("SF_CONF", "")
+
+        # Search the current directory, the directory under environment
+        # variable SF_CONF, and default user config directory, in that order.
+        directories = [os.getcwd(), self._env_config_dir, self._user_config_dir]
+        for directory in directories:
+            self._filepath = os.path.join(directory, self._name)
+            try:
+                self.load(self._filepath)
+            except FileNotFoundError:
+                # Check if we have searched all directories
+                if directory == directories[-1]:
```

I think this check could be avoided entirely if you modified the `load` method to return the config object at the end (which I think is good practice).

```python
directories = [os.getcwd(), self._env_config_dir, self._user_config_dir]
for directory in directories:
    self._filepath = os.path.join(directory, self._name)
    try:
        config = self.load(self._filepath)
        break
    except FileNotFoundError:
        pass

if config:
    self.update_config()
else:
    log.info("No Strawberry Fields configuration file found.")
```

---

This looks way better! Sad you can't yet do multi-line GitHub suggestions 😂

---

In this case, I would suggest that we throw and catch the exception in `load`, and return None if the file does not exist, to make this code cleaner. I.e. 

```
for directory in directories:
    self._filepath = os.path.join(directory, self._name)
    config = self.load(self._filepath)
    if config:
        self._config_file = config
        self.update_config()

```

---

The one disadvantage is that `load()` no longer will raise a `FileNotFoundError`, instead, it will silently fail. A developer building an app against SF is using `Configuration.load()` to manually load a configuration file will have to manually check that the function returned `None`, and manually implement error handling.

---

I implemented this as @zzzeid suggested but I agree with this, it depends on what the use case is. Do we intend to expose `load` as a method for developers to build against? If so, we should probably change this back. Or, we could change the method name to `_load`.

---

I think it's still an open question - to early to say how we intend for it to be used/how people might use it. My preference would be to change it back, as it could lead to unexpected or hard to debug behavior down the line.

---

My original comment was in relation to catching the exception and passing, as that would suppress the exception and not provide any indication as to what happened to the file loading. I think earlier we were catching the exception and showing a warning message. Essentially I guess we are still doing the same thing. I'm good with either implementation.

---

We'll always need to catch the exception and pass, this is more a question about where to do it. I feel it's better to do that in __init__ and not in load(), as (1) load could be developer/user facing, and (2) needing to pass on the exception is only required during initialization.

---

@josh146 This makes sense, and reduces the responsibility of the `load` method.

---

:+1: I'll make a PR to revert this behaviour.